contracts/upgradeable_contracts/ChaiConnector.sol

```diff
@@ -0,0 +1,278 @@
+        return rdivup(investedAmountInDai(), chi);
+    }
+
+    /**
+    * @dev Checks if DAI balance is high enough to be partially converted to Chai
+    * Twice limit is used in order to decrease frequency of convertDaiToChai calls,
+    * In case of high bridge utilization in DAI => xDAI direction,
+    * convertDaiToChai() will be called as soon as DAI balance reaches 2 * limit,
+    * limit DAI will be left as a buffer for future operations.
+    * @return true if convertDaiToChai() call is needed to be performed by the oracle
+    */
+    function isDaiNeedsToBeInvested() public view returns (bool) {
+        // chai token needs to be initialized, DAI balance should be at least twice greater than minDaiTokenBalance
+        return isChaiTokenEnabled() && daiBalance() > 2 * minDaiTokenBalance();
+    }
+
+    /**
+    * @dev Converts all DAI into Chai tokens, keeping minDaiTokenBalance() DAI as a buffer
+    */
+    function convertDaiToChai() public chaiTokenEnabled {
+        // there is not need to consider overflow when performing a + operation,
+        // since both values are controlled by the bridge and can't take extremely high values
+        uint256 amount = daiBalance().sub(minDaiTokenBalance());
+        setInvestedAmointInDai(investedAmountInDai() + amount);
+        erc20token().approve(chaiToken(), amount);
+        chaiToken().join(address(this), amount);
+
+        // When evaluating the amount of DAI kept in Chai using dsrBalance(), there are some fixed point truncations.
+        // The dependency between invested amount of DAI - value and returned value of dsrBalance() - res is the following:
+        // res = floor(floor(value / chi) * chi)), where chi is the coefficient from MakerDAO Pot contract
+        // This can lead up to losses of ceil(chi) DAI in this balance evaluation.
+        // The negative constant is needed here for making sure that everything works fine, and this error is small enough
+        require(dsrBalance() >= investedAmountInDai() - 10000);
+    }
+
+    /**
+    * @dev Redeems DAI from Chai, the total redeemed amount will be at least equal to specified amount
+    * @param amount Amount of DAI to redeem
+    */
+    function _convertChaiToDai(uint256 amount) internal {
+        uint256 invested = investedAmountInDai();
+        if (amount >= invested) {
+            // onExecuteMessage can call a convert operation with argument greater than the current invested amount,
+            // in this case bridge should withdraw all invested funds
+            chaiToken().draw(address(this), invested);
+            setInvestedAmointInDai(0);
+        } else if (amount > 0) {
+            uint256 initialDaiBalance = daiBalance();
+            chaiToken().draw(address(this), amount);
+            uint256 redeemed = daiBalance() - initialDaiBalance;
+            setInvestedAmointInDai(redeemed < invested ? invested - redeemed : 0);
+        }
+
+        require(dsrBalance() >= investedAmountInDai());
```

Comment:
Do you thinks that is more logical to check if the new balance of the bridge is as expected?

Reply:
What do you mean? 
Seems to me that if a new balance will not be as expected, we will get a revert anyway while trying to transfer DAI to user's account in `onExecuteMessage`.

Reply:
I meant that if you requested 300 dai (200 to execute transfer and 100 to have minimal balance) and by some reason receives only 250 dai as part of the execution `chaiToken().draw`, so we are in trouble.

Reply:
Ok, reasonable, I will add it.
What about a similar check in `convertDaiToChai`? We can also check that all approved amount was used.

Reply:
WRT `converDaiToChai`... Our main worry in this case is that all tokens we send will be considered as invested. So, correct me if I wrong: `chaiToken().join` performs an action as so the result of these actions will be reflected later in the result of `dsrBalance`. And the expected amount of invested tokens is calculated by `setInvestedAmointInDai(investedAmountInDai() + amount)`. If my reasoning is correct then  `require(dsrBalance() >= investedAmountInDai() - 10000)` checks that the expected amount not less than the actual amount of tokens. Since actions in `converDaiToChai` will not send more token than expected (`amount`) we don't need to check if the bridge balance was changed unexpectedly, do we?

But in `_convertChaiToDai` we could check both: expected amount of actual balance  and expected amount of invested tokens. In case if the expectation does not match -- revert the method call. Please share your view on this.

Reply:
Sounds correct. 
But what if something goes wrong in `converDaiToChai`, so that not all specified tokens are invested? For example, bridge approves `200` tokens, but Chai/MakerDAO contracts used only `150` of them, and `50` are still kept in approval. For such case, we also need to pay attention, don't we?

Reply:
they are still on the balance of the bridge so it is not the issue.